### ISSUE TITLE
What is the relationship between window/workDoneProgress/create and $/progress with kind=begin?

### ISSUE THREAD

**Issue by DanTup** [microsoft/language-server-protocol#1040]
I'm reading the 3.15 spec and I'm a little confused about how a server is supposed to report progress. There is a `window/workDoneProgress/create` request sent from the server to client but also a `$/progress` notification that supports a `{ kind: 'begin' }` structure.

If I want to start showing progress on the client, do I need to both call `window/workDoneProgress/create` and then immediately call `$/progress` with a `begin` event? If so, what is the reason for having to make two requests?

Similarly, when the progress is complete, do I need to send a `{kind: 'end'}` `$/progress` notification and then immediately call `window/workDoneProgress/cancel`?

**Comment by matklad** [comment]
>If so, what is the reason for having to make two requests?

I think I understand this. Each progress is reported against a token. A token can be supplied by the client (when sending a request) or by the server (when the server initiates a new progress, unrelated to any request). `window/workDoneProgress/create` registers server token. 

It would have been possible to get by without `window/workDoneProgress/create`, by using the following logic on the client side for handling `{kind: 'end'}`:

* If there's an outstanding client progress token, associated this report with it
* otherwise, *assume* that this is a server-initiated progress and create a new client-side progress resource. 

The problem with this is that it becomes harder to spot bugs in the server, when it accidentally reports the progress against a wrong token (ie, using a token for an already disposed request). 

>do I need to send a {kind: 'end'} $/progress notification and then immediately call window/workDoneProgress/cancel?

In rust-analyzer, we issue only notification, and that seems to work well. THough, I didn't know that `window/workDoneProgress/cancel` exists :)

**Comment by DanTup** [comment]
>  it becomes harder to spot bugs in the server, when it accidentally reports the progress against a wrong token (ie, using a token for an already disposed request).

I can see some value in that, but now the server has an extra round trip. It seems like it could've easly been included in the `begin` progress notification (eg. as a different `kind` or with a flag saying it is a server-created progress notification and doesnt' correspond to a client request).

So I guess the answer is Yes, the server needs to send two requests to begin (and end) a server-initiated progress? Probably the spec should make this a bit more explicit (since on the surface it feels a bit strange IMO).

**Comment by matklad** [comment]
>I can see some value in that, but now the server has an extra round trip.

Hm, good point, in rust-analyzer I think we are actually non-conforming, as we start sending the progress without waiting for the response to arrive... Properly awaiting response would be awkward, as you'd have to queue progress events which arrived in the meantime...

**Comment by alanz** [comment]
There was some earlier discussion around this (and a quick search failed me), where the answer was something like "the initial request can potentially allocate UI resources on the server, and may fail", hence it is a request/response not a notify.

The upshot being that the token represents actual UI resources on the client, and so needs to be properly managed from the server, to allocate and then free it at the end.

**Comment by DanTup** [comment]
> in rust-analyzer I think we are actually non-conforming, as we start sending the progress without waiting for the response to arrive...

Yeah, it's possible clients will show errors to the user if they get notifications for progresses that they failed to create.

> Properly awaiting response would be awkward, as you'd have to queue progress events which arrived in the meantime...

Yep. Either that or you wait for the response before starting the work, but that also feels a bit silly.

> The upshot being that the token represents actual UI resources on the client, and so needs to be properly managed from the server, to allocate and then free it at the end.

This feels kinda weird. In another issue it was mentioned that LSP servers [should not be aware of UI state](https://github.com/microsoft/language-server-protocol/issues/1031#issuecomment-648706536) but here it seems like the server is specifically having to manage UI resources for the editor.

**Comment by rwols** [comment]
> If I want to start showing progress on the client, do I need to both call window/workDoneProgress/create and then immediately call $/progress with a begin event?

From what I understand (and also from what is implemented in the sublime text client), yes.

> If so, what is the reason for having to make two requests?

`window/workDoneProgress/create` is a request. `$/progress` is a notification.

I don't know the underlying reason. From an implementation point of view it could have just as well been a single `$/progress` with kind `begin`. But I can only speak for the sublime text client.

Currently, we use `window/workDoneProgress/create` to allocate data for reporting. Once the data structures are ready we respond with `null`. [It's a bit silly because all we do is add a dict to a map somewhere](https://github.com/sublimelsp/LSP/blob/e7dac40247d68398fc24d1b19b1565d05e472ae8/plugin/core/windows.py#L585-L605).

> Similarly, when the progress is complete, do I need to send a {kind: 'end'} $/progress notification and then immediately call window/workDoneProgress/cancel?

No, only `$/progress` with kind `end` should be sufficient. It is *this* notification that we use to de-allocate data for a progress report. You could let a datastructure grow arbitrarily large by not sending progress notifications of kind `end`.

**Comment by dbaeumer** [comment]
@alanz pointed out correctly the create exists because the server allocates UI resources on the client. In case of client initiated progress the client already allocated the UI resource and manages it.

`window/workDoneProgress/cancel` is a notification send from the client to the server it indicate that the user has pressed the Cancel button for the server initiated progress. For client initiated progress this is not necessary since every request already carries a cancelation token (e.g. `$/cancelRequest` with the request id).

Any objection to mark this issue as discussion and close it?

**Comment by DanTup** [comment]
> `window/workDoneProgress/cancel` is a notification send from the client 

Oops, I misread that and thought the server was needing to send it.

> Any objection to mark this issue as discussion and close it?

That's fine - I think everything is clear now. It does feel a little strange that the server must wait for an async response before sending progress though (esp if the client is busy/single-threaded), but I guess changing how this works now is breaking and unlikely to happen. I'll just implement something that delays the progress notifications until the client has responded.

Thanks!
